### DISCUSSION TITLE
Hostname scheme and port not picked up correctly?

### DISCUSSION THREAD

**Discussion by akiross** [keycloak/keycloak#33490] [Q&A] [answered]
Hello,

I'm trying to set up KC (24.0.5) for production on a public server: ports 80 and 443 are exposed to the public (say `auth.example.com`), all the other ports are closed to the internet, but I can still access all the ports via VPN host (say 192.168.10.10).

I can get KC running inside the VPN: `hostname` is set to 192.168.10.10 and `https-port` is set to 8443: I can access https://192.168.10.10:8443 from my browser and the login console is loaded correctly, I can login as admin and so on.

Now I want to expose it to the world, but there's a proxy in between (caddy). I would like to have the proxy forwarding connections from 443 to port 8443 and have keycloak reachable via its public name.

So, I keep `https-port` set to 8443, but I change the hostname to the public facing one: `hostname` is `auth.example.com`; I guess that's bound to fail, since the public facing port is not 8443, but 443, but bear with me.
When I reload my config and go to `https://auth.example.com`, the UI starts to load, but a specific call will fail:
```
https://auth.example.com:8443/realms/master/protocol/openid-connect/3p-cookies/step1.html
```

If I adjust the hostname to include the port (`hostname` set to `auth.example.com:443`), when I load the `https://auth.example.com` page, all the requests are now successful, but the page loads forever (the circle spinner goes on and on).
I opened the HTTP page with the spinner, and there's this snippet that look very suspicious:

```

  <script id="environment" type="application/json">
    {
      "loginRealm": "master",
      "clientId": "security-admin-console",
      "authServerUrl": "https://auth.example.com:443:8443", ////////////////////////////////////////////// Why?
      "authUrl": "https://auth.example.com",
      "consoleBaseUrl": "/admin/master/console/",
      "resourceUrl": "/resources/6pls6/admin/keycloak.v2",
      "masterRealm": "master",
      "resourceVersion": "6pls6",
      "logo": "",
      "logoUrl": ""
    }
  </script>
```

If I also add `https://` in front of the hostname (`hostname` set to `https://auth.example.com:443`), then loading the public page will fail and the request above will still fail, but with this url:

```
https://https//auth.example.com:443:8443/realms/master/protocol/openid-connect/3p-cookies/step1.html
```

Something similar happens if I exclude the port, like `https://auth.example.com`: a request is performed for the URL
```
https://https//auth.example.com:8443/realms/master/protocol/openid-connect/3p-cookies/step1.html
```

From [the docs](https://www.keycloak.org/server/hostname#_defining_specific_parts_of_the_hostname_option), it seems that KC should pick up the URL and use that one, possibly parsing the various parts and fixing what has to be fixed, but it doesn't seem like it.

I'm using a configuration file like this:

```
hostname=https://auth.example.com:443
https-port=8443
```

Am I mis-reading the doc or is this some kind of bug?

**Comment by jonkoops** [comment, marked as answer]
The issues you are experiencing may have already been fixed in a newer version of Keycloak. Have you tried running the latest 25.x release? It [introduced some changes](https://www.keycloak.org/2024/06/keycloak-2500-released#_new_hostname_options) to the hostname configuration to reduce common issues like this.

**Reply by akiross** [reply]
Ok, I realized a bit too late v25 was out: I'll try with that and let you know.

**Reply by akiross** [reply]
Indeed, v25 worked as expected.

**Comment by shawkins** [comment]
> Am I mis-reading the doc or is this some kind of bug?

Those docs are specific to KC 25+. For KC 24 you would use https://www.keycloak.org/server/hostname-deprecated

If possible can you retry your setup with keycloak 25 using the new docs you are referencing for hostname v2?

**Reply by akiross** [reply]
I will try with v25 and report back, thanks

**Reply by akiross** [reply]
I tried with v25 and it works, thank you! From the doc I couldn't understand when to use v1 or v2

**Reply by shawkins** [reply]
The version for the docs is seen on https://www.keycloak.org/documentation - along with links to archived versions. However you are correct that each doc page does not clearly show which version it it associated with. Can you log an enhancement request for whatever would make this clearer?

**Reply by shawkins** [reply]
Captured the doc versioning issue as #33532
